File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int answer= 0x7fffffff;
string targetAns;

map<string, bool> m;

bool stringEqual(string a, string b) {
    int tmp = 0;
    for(int i=0; i<a.length(); i++) {
        if(a[i] != b[i]) tmp++;
    }
    if(tmp >= 2) return false;
    return true;
}

void backtrack(string s, vector<string> words, int tmp) { // hit
    
    if(s == targetAns) {
        answer = min(answer, tmp);
        return;
    }
    
    if(tmp > words.size()) return;
    
    for(auto w: words) {
        if(stringEqual(s, w) == true && m[w] == false) {
            m[w] = true;
            backtrack(w, words, tmp+1);
            m[w] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {
    targetAns = target;
    for(auto w : words) {
        m[w] = false;
    }
    backtrack(begin, words, 0);
    if(answer == 0x7fffffff) return 0;
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    
    int ans = Integer.MAX_VALUE;
  
    public int solution(String begin, String target, String[] words) {
        for(int i=0; i<words.length; i++) {

            int diff = 0;
            for(int j=0; j<words[i].length(); j++) {
                if(words[i].charAt(j) != begin.charAt(j)) {
                    diff++;
                }
            }
            
            if(diff == 1) {
                boolean[] visited = new boolean[words.length];
                visited[i] = true;
                dfs(words[i], visited, 1, words, target);
             }
        }
        if(ans == Integer.MAX_VALUE) return 0;
        else return ans;
    }
    
    public void dfs(String cur, boolean[] visited, int cnt, String[] words, String target) {
        if(cur.equals(target)) {
            ans = Math.min(ans, cnt);
            return;
        }
        
        for(int i=0; i<words.length; i++) {
            if(visited[i]) continue;

            int diff = 0;
            for(int j=0; j<words[i].length(); j++) {
                if(words[i].charAt(j) != cur.charAt(j)) {
                    diff++;
                }
            }
            
            if(diff == 1) {
                visited[i] = true;
                dfs(words[i], visited, cnt+1, words, target);
                visited[i] = false;
             }
        }
       
    }
    
}